### PR TITLE
Better fullscreen

### LINKED ISSUE #571: Fix fullscreen if there is no current webview (zim)
The fullscreen functionality works by putting the current (zim) webview fullscreen.
But there is no safeguard if there is no current zim displayed and we end by passing a nullptr to the FullScreen class.

We have a technical bug (using a nullptr) and a functional point to fix (what to do if there is no current webview)

The simpler seems to deactivate the fullscreen if there is no current zim.
But then we have to deactivate the action, not just add a safeguard.
And we probably need to test corner use case (closing the webview while we are fullscreen, ...)

Or we may want to keep the fullscreen functionality if there is no current zim, but we must decide what is the correct behavior.

### PR DESCRIPTION
**Pull request by juuz0**
Fixes #571

### CODE REVIEW

**Review by kelson42 — CHANGES_REQUESTED** [review]
@juuz0 Feature works fine, but seems to me incomplete. Would you be able to show full window (but without window decoration) when mouse in on top of screen like on Firefox? This is important because otherwise we might have easily users stuck in fullscreen.

**Comment by juuz0** [comment]
> This is important because otherwise we might have easily users stuck in fullscreen.

Agreed, haha. Somehow I missed implementing that. On it.

File: src/tabbar.h [hunk 1]
```diff
@@ -53,6 +53,7 @@ class TabBar : public QTabBar
     void webActionEnabledChanged(QWebEnginePage::WebAction action, bool enabled);
     void libraryPageDisplayed(bool displayed);
     void currentTitleChanged(const QString& title);
+    void toggleFullscreen();
```
[the remark this answers is not in the document]
Comment (juuz0) [reply src/tabbar.h:56]: Overlooked it, thanks. Will fix it

**Comment by kelson42** [comment]
@juuz0 Any chance this PR could go through next review iteration before the end of the week? Otherwise I will have probably to move it to milestone 2.3.0.

**Comment by juuz0** [comment]
@kelson42 Yes, in the next 24 hours :+1:

**Comment by juuz0** [comment]
>  Would you be able to show full window (but without window decoration) when mouse in on top of screen like on Firefox? 

Done, also rebased.

You can review but I want to ask - should the ZIM's fullscreen (but don't keep window decoration) be removed on pressing the fulltext search command shortcut (Ctrl L, for me) and then after search is done, back to fullscreen.

**Comment by juuz0** [comment]
Whoa, what happened to CI

**Review by kelson42 — CHANGES_REQUESTED** [review]
@juuz0 This does not behave like it should, tobar/tabs headers appears now when you put your mouse on the top of the screen but:
* They don't disappear after a while
* It seems you actualy leave the fulltscreen mode, but you still does not display the window decoration.
* There is somehow a flash

This is not like it work for Firefox for example, look at it to better understand what is expected please.

**Comment by kelson42** [comment]
@juuz0 What should we do with this PR?

**Comment by juuz0** [comment]
@kelson42  I will re-review this and let you know by today :+1:

**Comment by kelson42** [comment]
@juuz0 Any news here?

**Comment by juuz0** [comment]
@kelson42 I forgot most of what I did earlier, but now I've re reviewed :+1: Thanks for the ping

Curently, what I am doing is:

- Put current webview in fullscreenwindow
- When mouse is moved to top of screen, remove fullscreenwindow and put MainScreen in full screen (this change is the cause of a "flash") - this is to mimic Firefox's behaviour which shows tabs on bringing mouse to top of screen.

What this lacks:
- [in Firefox] When mouse is brought to top of screen and tab is changed by mouse click, the new tab is also made fullscreen. Doesn't happen with this PR, I've tried implementing but it becomes a mesh with flashes. Not good.
-  Turn back to fullscreen mode when mouse is not on top of screen. Easy to do but flashes will become a problem.
- Firefox's minimize, maximize, exit buttons are on the same line as tabs, here we don't. So how could they be shown? Or we don't show them?

The current implementation doesn't look right.
My thinking on Firefox's implementation is that it puts the application in fullscreen and then hides/unhides the tabs part, maybe that should be done. This red part will get toggled for visibility.
image

Do you have some other way to implement this? @mgautierfr

**Comment by juuz0** [comment]
Closing in favour of #859
